### PR TITLE
Bug 763020 and Bug 763021 - sync subset of engines and do a clients only sync after Send Tab

### PR DESCRIPTION
**Pull request by ncalexan**
I spent a few days implementing a data driven sync, computing dependencies with a topological sort, etc, etc, and when I was done I looked at the patch set and decided that it was a lot of churn for not a lot of win.  Our dependency tree is a line with a diamond of ServerSyncStages (with no interdependencies) in the middle. I think the approach of this ticket, which just has each stage check to see if it is disabled (like we already do if the stage is not present in meta/global) supports this dependency graph well.

See 763021 for detailed device testing instructions.

### CODE REVIEW

**Comment by ncalexan** [comment]
rnewman, perhaps you should look at this, since we've discussed this quite a bit.

File: src/main/java/org/mozilla/gecko/sync/GlobalSession.java [hunk 1]
```diff
@@ -162,6 +162,11 @@ public GlobalSession(String userAPI,
 
     registerCommands();
     prepareStages();
+    Collection<String> knownStageNames = new HashSet<String>();
+    for (Stage stage : Stage.getNamedStages()) {
+      knownStageNames.add(stage.getRepositoryName());
+    }
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/GlobalSession.java:168]: Seems kinda redundant to compute this unnecessarily. Why not have `getStagesToSyncFromBundle` return `null` and do this conditionally?

Or, more cleanly, move `getStagesToSync*` into `Stage`, so it knows about the named stages, and you can avoid some of the generality and complexity.

File: src/main/java/org/mozilla/gecko/sync/SyncConfiguration.java [hunk 2]
```diff
@@ -195,7 +196,16 @@ public void unregisterOnSharedPreferenceChangeListener(OnSharedPreferenceChangeL
...
-  public Set<String>     enabledEngineNames;
+  public Set<String> enabledEngineNames;
+
+  /**
+   * Names of stages to sync <it>this sync</it>, or <code>null</code> to sync
+   * all known stages.
+   * <p>
+   * Generated <it>each sync</it> from extras bundle passed to
+   * <code>SyncAdapter.onPerformSync</code> and not persisted.
+   */
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/SyncConfiguration.java]: Add a note about not being synchronized…

File: src/main/java/org/mozilla/gecko/sync/setup/Constants.java [hunk 3]
```diff
@@ -18,6 +18,24 @@
...
   public static final String DATA_ENABLE_ON_UPGRADE = "data.enableOnUpgrade";
 
+  /**
+   * Key in sync extras bundle specifying stages to sync this sync session.
+   * <p>
+   * Corresponding value should be a String JSON-encoding an object, the keys of
+   * which are the stage names to sync. For example:
+   * <code>"{ \"stageToSync\": 0 }"</code>.
+   */
+  public static final String EXTRAS_KEY_STAGES_TO_SYNC = "stagesToSync";
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/setup/Constants.java]: What's wrong with "sync" and "skip"?

File: src/main/java/org/mozilla/gecko/sync/syncadapter/SyncAdapter.java [hunk 4]
```diff
@@ -221,6 +223,29 @@ public boolean shouldBackOff() {
...
+   */
+  public static void requestImmediateSync(final Account account, final String[] stageNames) {
+    if (account == null) {
+      Logger.warn(LOG_TAG, "Not requesting immediate sync because Android Account is null.");
+      return;
+    }
+
+    final Bundle extras = new Bundle();
+    Utils.putStageNamesToSync(extras, stageNames, null);
+    extras.putBoolean(ContentResolver.SYNC_EXTRAS_MANUAL, true);
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/syncadapter/SyncAdapter.java:245]: We check for "force" somewhere else in the code (skipping backoff ourselves…?). Clean that up, too?

File: src/main/java/org/mozilla/gecko/sync/syncadapter/SyncAdapter.java [hunk 5]
```diff
@@ -234,6 +259,10 @@ public void onPerformSync(final Account account,
     this.syncResult   = syncResult;
     this.localAccount = account;
 
+    Log.i(LOG_TAG, "Syncing client named " + getClientName() +
+                   " with client guid " + getAccountGUID() +
+                   " (sync account has " + getClientsCount() + " clients).");
+
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/syncadapter/SyncAdapter.java]: Did you leave this here deliberately?

File: test/src/org/mozilla/gecko/sync/test/TestUtils.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
+package org.mozilla.gecko.sync.test;
```
Comment (rnewman) [inline test/src/org/mozilla/gecko/sync/test/TestUtils.java:4]: License.

File: test/src/org/mozilla/gecko/sync/test/TestUtils.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.mozilla.android.sync.test.AndroidSyncTestCase;
+import org.mozilla.gecko.sync.Utils;
+
+import android.os.Bundle;
+
+public class TestUtils extends AndroidSyncTestCase {
```
Comment (rnewman) [inline test/src/org/mozilla/gecko/sync/test/TestUtils.java:18]: `TestStageComputation`?

**Comment by rnewman** [comment]
One more go 'round. Looking good.

**Comment by ncalexan** [comment]
liuche:  I renamed `s1` and `s2`.  No tests because I didn't feel like re-working existing code to support testing.  The STRs on the ticket are pretty clear.  If you need to see this tested I'll re-work things.

**Pull request merged**